### PR TITLE
Cached assets attempt 1

### PR DESCRIPTION
**Pull request by rowanseymour**
Reworks asset representation so that we can model embedded and lazy assets

```json
{"type": "flow", "content": {}, "fetch_url": "http://rpd.io/assets/flow/1234..."}
```

```json
{"type": "flow", "uuid": "1234...", "fetch_url": "http://rpd.io/assets/flow/1234..."}
```

Haven't yet implemented fetching of lazy assets

### CODE REVIEW

File: flows/engine/assets.go [hunk 1]
```diff
@@ -3,90 +3,142 @@ package engine
...
 	"github.com/nyaruka/goflow/utils"
 )
 
-type assets struct {
-	flows    map[flows.FlowUUID]flows.Flow
-	channels map[flows.ChannelUUID]flows.Channel
+type assetContainer struct {
+	assetType  flows.AssetType
+	asset      flows.Asset
+	accessedOn *time.Time
```
Comment (rowanseymour) [inline flows/engine/assets.go]: I'm thinking we'll eventually evict stuff from the asset cache based on when it was last accessed

File: flows/engine/assets.go [hunk 2]
```diff
@@ -3,90 +3,142 @@ package engine
...
+			m.AddLazyAsset(envelope.Type, *envelope.UUID, envelope.URL)
 		}
-		channelList[c] = channel
 	}
 
-	for f := range envelope.Flows {
-		flow, err := definition.ReadFlow(envelope.Flows[f])
-		if err != nil {
-			return nil, err
+	// any non-lazy assets can be validated now
```
Comment (rowanseymour) [inline flows/engine/assets.go]: This is a bit weird - say someone passes in a flow and a subflow which the first flow requires. If we validate each asset before adding it, the first one will trigger another request for the subflow... so I'm adding them all to the asset cache and then validating them.

File: cmd/flowserver/main.go [hunk 3]
```diff
@@ -33,6 +37,7 @@ import (
 )
 
 var version = "dev"
+var assetManager flows.AssetManager
```
Comment (rowanseymour) [inline cmd/flowserver/main.go]: This asset manager will exist for the lifetime of the server and be shared by all requests

File: flows/engine/assets.go [hunk 4]
```diff
@@ -3,90 +3,147 @@ package engine
...
+	fetchURL   string
 }
 
-func NewAssets(flowList []flows.Flow, channelList []flows.Channel) flows.Assets {
-	flowMap := make(map[flows.FlowUUID]flows.Flow, len(channelList))
-	for _, flow := range flowList {
-		flowMap[flow.UUID()] = flow
+type assetManager struct {
+	cache      map[flows.AssetUUID]assetContainer
+	cacheMutex sync.Mutex
```
Comment (nicpottier) [inline flows/engine/assets.go]: I think you can use a `RWMutex` here and it'll operate a bit faster for the read cases.

File: flows/engine/assets.go [hunk 5]
```diff
@@ -3,90 +3,147 @@ package engine
...
+}
+
+func NewAssetManager() flows.AssetManager {
```
Comment (nicpottier) [inline flows/engine/assets.go]: Not passionate, but I think I prefer just `Assets` over `AssetManager`.. `Manager` always smells a bit as to me it doesn't carry a lot of meaning with it.

**Comment by nicpottier** [comment]
Your comment above says `fetch_url` but it seems we use `url` is that right? (I prefer `url`)

File: flows/engine/assets.go (hunk 5 above)
Comment (rowanseymour) [inline flows/engine/assets.go]: AssetStore?

**Comment by rowanseymour** [comment]
@nicpottier yeah I changed it to just `url` at some point

**Pull request merged**